crypto/s2n_ecc_evp.c

```diff
@@ -151,6 +149,131 @@ int s2n_ecc_evp_compute_shared_secret_from_params(struct s2n_ecc_evp_params *pri
+    size_t ret = EC_POINT_point2oct(group, point, POINT_CONVERSION_UNCOMPRESSED, NULL, 0, NULL);
+    S2N_ERROR_IF(ret == 0, S2N_ERR_ECDHE_SERIALIZING);
+    S2N_ERROR_IF(ret > UINT8_MAX, S2N_ERR_ECDHE_SERIALIZING);
+    *length = (uint8_t)ret;
+    return 0;
+}
+
+static int s2n_ecc_evp_write_point_data_snug(const EC_POINT *point, const EC_GROUP *group, struct s2n_blob *out) {
+    size_t ret = EC_POINT_point2oct(group, point, POINT_CONVERSION_UNCOMPRESSED, out->data, out->size, NULL);
+    S2N_ERROR_IF(ret != out->size, S2N_ERR_ECDHE_SERIALIZING);
+    return 0;
+}
+
+static EC_POINT *s2n_ecc_evp_blob_to_point(struct s2n_blob *blob, const EC_KEY *ec_key) {
+    const EC_GROUP *group = EC_KEY_get0_group(ec_key);
+    EC_POINT *point = EC_POINT_new(group);
+    if (point == NULL) {
+        S2N_ERROR_PTR(S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+    }
+    if (EC_POINT_oct2point(group, point, blob->data, blob->size, NULL) != 1) {
+        EC_POINT_free(point);
+        S2N_ERROR_PTR(S2N_ERR_BAD_MESSAGE);
+    }
+    return point;
+}
+#endif
+
+int s2n_ecc_evp_read_params_point(struct s2n_stuffer *in, int point_size, struct s2n_blob *point_blob) {
+    notnull_check(in);
+    notnull_check(point_blob);
+    gte_check(point_size, 0);
+
+    /* Extract point from stuffer */
+    point_blob->size = point_size;
+    point_blob->data = (uint8_t *)s2n_stuffer_raw_read(in, point_size);
+    notnull_check(point_blob->data);
+
+    return 0;
+}
+
+int s2n_ecc_evp_write_params_point(struct s2n_ecc_evp_params *ecc_evp_params, struct s2n_stuffer *out) {
+    notnull_check(ecc_evp_params);
+    notnull_check(ecc_evp_params->negotiated_curve);
+    notnull_check(ecc_evp_params->evp_pkey);
+    notnull_check(out);
+
+#if S2N_IS_X25519_SUPPORTED
+    out->blob.size = EVP_PKEY_get1_tls_encodedpoint(ecc_evp_params->evp_pkey, &out->blob.data);
+    S2N_ERROR_IF(out->blob.size != ecc_evp_params->negotiated_curve->share_size, S2N_ERR_ECDHE_SERIALIZING);
+
+    out->blob.data = s2n_stuffer_raw_write(out, out->blob.size);
+    notnull_check(out->blob.data);
+#else
+    uint8_t point_len;
+    DEFER_CLEANUP(EC_KEY *ec_key = EVP_PKEY_get1_EC_KEY(ecc_evp_params->evp_pkey), EC_KEY_free_pointer);
+    S2N_ERROR_IF(ec_key == NULL, S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+
+    const EC_POINT *point = EC_KEY_get0_public_key(ec_key);
+    const EC_GROUP *group = EC_KEY_get0_group(ec_key);
+    S2N_ERROR_IF(point == NULL || group == NULL, S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+
+    GUARD(s2n_ecc_evp_calculate_point_length(point, group, &point_len));
+    S2N_ERROR_IF(point_len != ecc_evp_params->negotiated_curve->share_size, S2N_ERR_ECDHE_SERIALIZING);
+    out->blob.data = s2n_stuffer_raw_write(out, point_len);
+    out->blob.size = point_len;
+
+    notnull_check(out->blob.data);
+    GUARD(s2n_ecc_evp_write_point_data_snug(point, group, &out->blob));
+#endif
+    return 0;
+}
+
+int s2n_ecc_evp_parse_params_point(struct s2n_blob *point_blob, struct s2n_ecc_evp_params *ecc_evp_params) {
```

Comment:
Do we need to validate the point?

Reply:
Yes, this is currently being used in the client_key_share.c here: https://github.com/awslabs/s2n/blob/56cd7a9e6400e298ff83cf767766d30a49107b5b/tls/extensions/s2n_client_key_share.c#L117 and server key_share.c https://github.com/awslabs/s2n/blob/56cd7a9e6400e298ff83cf767766d30a49107b5b/tls/extensions/s2n_server_key_share.c#L147

Reply:
From https://tools.ietf.org/html/rfc8446#page-51: 

```
 This process consists of three steps: 
(1) verify that Q is not the point at infinity (O), 
(2) verify that for Q = (x, y) both integers x and y are in the correct
   interval, and 
(3) ensure that (x, y) is a correct solution to the
   elliptic curve equation.  

For these curves, implementors do not need
   to verify membership in the correct subgroup.

```

This is valid for both TLS1.2 and TLS1.3. Use the OpenSSL function `EC_KEY_check_key` to check validity of the keys.